### DISCUSSION TITLE
laravel 7 blade component public methods not working

### DISCUSSION THREAD

**Discussion by mohammadmojrian** [laravel/framework#33454] [General]
I have a problem with my Laravel project in using blade component.

When I define a new component and want to use "Component public Methods" to pass data to "view component blade file", I receive an error, "Undefined variable"

The component has been made by

`php artisan make:component testc 
`
has following codes:
**component code**

```
<?php

namespace App\View\Components;

use Illuminate\View\Component;

class testc extends Component
{
    /**
     * Create a new component instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Get the view / contents that represent the component.
     *
     * @return \Illuminate\View\View|string
     */
    public function render()
    {
        return view('components.testc');
    }

    public function test()
    {
        return 'test';
    }
}
```

**view component blade file**

```
<div>
    {{ $test }}
</div>
```

**Error**

```
Facade\Ignition\Exceptions\ViewException
Undefined variable: test (View: /../resources/views/components/testc.blade.php)
```

i try `cache:clear` `view:clear` and `composer dump` before

laravel version : **7.18.0**
php version : **7.2.24**

**Comment by driesvints** [comment]
Your testc class name is wrong. It needs to start with an uppercase letter as the PSR-4 autoloading standard requires:

```php
<?php

namespace App\View\Components;

use Illuminate\View\Component;

class Testc extends Component
```

**Reply by mohammadmojrian** [reply]
thanks for your answer 
but my component work correctly 
and i edit my file name and class name to pass PSR-4 Standard 
I have the same problem to pass any public methods to components blade file

**Comment by driesvints** [comment]
I tried out your code and everything works fine for me:

```
            <div class="content">
                <div class="title m-b-md">
                    Laravel
                </div>

                <x-testc />
```

![Screenshot 2020-07-07 at 14 41 39](https://user-images.githubusercontent.com/594614/86782935-fb5fb100-c05f-11ea-876e-d3e8352d9b08.png)

**Reply by mohammadmojrian** [reply]
thanks @driesvints ,

i change my tag from `<x-testc></x-testc>` to self closing tag `<x-testc />` it's work thanks

**Reply by mohammadmojrian** [reply]
@driesvints
after create new component i have problem again :(
i create a new component /section/index/features and create a public methos to return array
in blade change code to loop a variable ( my function ) and try to print 
use <x-section.index.features></x-section.index.features> and <x-section-index.features> not working agian :(
i thinks this conflicted with any bug or something...
because in my other projects is working well

**Reply by driesvints** [reply]
It works fine for me like `<x-testc></x-testc>` as well.
